File: typescript/src/Part1. Introduction To Functional Programming/3. Functional Data Structures/tree.ts
import { absurd } from "fp-ts/lib/function"

// 57p
class Leaf<A> {
  readonly _tag = "Leaf"
  constructor(readonly value: A) { }
}

class Branch<A> {
  readonly _tag = "Branch"
  constructor(readonly left: Tree<A>, readonly right: Tree<A>) { }
}

export type Tree<A> = Leaf<A> | Branch<A>

export const leaf = <A>(value: A): Tree<A> => new Leaf(value)
export const branch = <A>(left: Tree<A>, right: Tree<A>): Tree<A> =>
  new Branch(left, right)

interface Match<A, B> {
  Leaf: (v: A) => B,
  Branch: (l: Tree<A>, r: Tree<A>) => B
}

export const match = <A>(tree: Tree<A>) => <B>(m: Match<A, B>): B => {
  switch (tree._tag) {
    case 'Leaf':
      return m.Leaf(tree.value)
    case 'Branch':
      return m.Branch(tree.left, tree.right)
    default:
      return absurd(tree)
  }
}

// Exercise 3-25
export const size = <A>(tree: Tree<A>): number => {
  return match(tree)({
    'Branch': (l, r) => 1 + size(l) + size(r),
    'Leaf': () => 1
  })
}

// Exercise 3-26
export const maximum = (tree: Tree<number>): number => {
  return match(tree)({
    'Branch': (l, r) => Math.max(maximum(r), maximum(l)),
    'Leaf': (v) => v
  })
}

// Exercise 3-27
export const depth = <A>(tree: Tree<A>): number => {
  return match(tree)({
    'Branch': (l, r) => 1 + depth(r) + depth(l),
    'Leaf': () => 1
  })
}

// Exercise 3-28
export const map = <A, B>(f: (a: A) => B) => (tree: Tree<A>): Tree<B> => {
  return match(tree)({
    'Branch': (l, r) => branch(map(f)(l), map(f)(r)),
    'Leaf': (v) => leaf(f(v))
  })
}

// TODO: Exercise 3-29

File: scala/fpinscala/src/main/scala/fpinscala/exercises/datastructures/Tree.scala
package fpinscala.exercises.datastructures

sealed trait Tree[+A]
case class Leaf[A](value: A) extends Tree[A]
case class Branch[A](left: Tree[A], right: Tree[A]) extends Tree[A]

object Tree {
  // exercises 3-25
  def size[A](t: Tree[A]): Int =
    t match {
      case Leaf(_)             => 1
      case Branch(left, right) => size(left) + size(right)
    }

  // exercises 3-26
  def maximum(t: Tree[Int]): Int =
    t match {
      case Leaf(value)         => value
      case Branch(left, right) => maximum(left) max maximum(right)
    }

  // exercises 3-27
  def depth[A](t: Tree[A]): Int =
    t match {
      case Leaf(_)             => 0
      case Branch(left, right) => 1 + (depth(left) max depth(right))
    }

  // exercises 3-28
  def map[A, B](t: Tree[A])(f: A => B): Tree[B] =
    t match {
      case Leaf(value)         => Leaf(f(value))
      case Branch(left, right) => Branch(map(left)(f), map(right)(f))
    }

  // TODO: exercises 3-29
  // @see https://github.com/fpinscala/fpinscala/blob/da183734af0ebbc1c0ae8fa4bd9e8ec5acbd1216/answers/src/main/scala/fpinscala/datastructures/Tree.scala#L40
  def fold[A, B](t: Tree[A])(f: A => B)(g: (B, B) => B): B =
    t match {
      case Leaf(value)         => f(value)
      case Branch(left, right) => g(fold(left)(f)(g), fold(right)(f)(g))
    }

  def size2[A](t: Tree[A]): Int =
    fold(t)(_ => 1)(_ + _)

  def maximum2(t: Tree[Int]): Int = {
    def id[A](a: A): A = a

    fold(t)(id)(_ max _)
  }

  def depth2[A](t: Tree[A]): Int =
    fold(t)(_ => 0)((d1, d2) => 1 + (d1 max d2))

  def map2[A, B](t: Tree[A])(f: A => B): Tree[B] =
    fold(t)(a => Leaf(f(a)): Tree[B])(Branch(_, _))
}
